In [1]:
import pandas as pd
import numpy as np
ctd = pd.read_csv("../Data/downcast_all.csv")
print(ctd.head())

/var/folders/xn/9gqx3sxx3s32k1ttvhjc4fch0000gn/T/ipykernel_49220/975230851.py:3: DtypeWarning: Columns (2,3,7,8,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83) have mixed types. Specify dtype option on import or set low_memory=False.
  ctd = pd.read_csv("../Data/downcast_all.csv")


   PROJECT   STUDY ORD_OCC EVENT_NUM    CAST_ID         DATE_TIME_UTC  \
0  CalCOFI  9308NH       1       NaN  9308_001d  1993-08-11T12:05:46Z   
1  CalCOFI  9308NH       1       NaN  9308_001d  1993-08-11T12:06:04Z   
2  CalCOFI  9308NH       1       NaN  9308_001d  1993-08-11T12:06:10Z   
3  CalCOFI  9308NH       1       NaN  9308_001d  1993-08-11T12:06:11Z   
4  CalCOFI  9308NH       1       NaN  9308_001d  1993-08-11T12:06:14Z   

          DATE_TIME_PST LAT_DEC LON_DEC       STA_ID  ...  OXBUM CHL_A PHAEO  \
0  1993-08-11T04:05:46Z     NaN     NaN  093.3 026.7  ...    NaN   NaN   NaN   
1  1993-08-11T04:06:04Z   -99.0   -99.0  093.3 026.7  ...  242.7  0.18  0.05   
2  1993-08-11T04:06:10Z   -99.0   -99.0  093.3 026.7  ...    NaN   NaN   NaN   
3  1993-08-11T04:06:11Z   -99.0   -99.0  093.3 026.7  ...  243.9  0.16  0.05   
4  1993-08-11T04:06:14Z   -99.0   -99.0  093.3 026.7  ...    NaN   NaN   NaN   

   NO3  NO2  NH4  PO4  SIL FINAL_FLAG CAST_COUNT  
0  NaN  NaN  NaN  NaN  NaN   

In [5]:
print(ctd.columns)
print(ctd.tail())

Index(['PROJECT', 'STUDY', 'ORD_OCC', 'EVENT_NUM', 'CAST_ID', 'DATE_TIME_UTC',
       'DATE_TIME_PST', 'LAT_DEC', 'LON_DEC', 'STA_ID', 'LINE', 'STA', 'DEPTH',
       'PRESSURE', 'PRQ', 'TEMP1', 'TEMP1Q', 'TEMP2', 'TEMP2Q', 'TEMPAVE',
       'SALT1', 'SALT1Q', 'SALT1_CORR', 'SALT2', 'SALT2Q', 'SALT2_CORR',
       'SALTAVE_CORR', 'OX1', 'OX1Q', 'OX1_CRUISECORR', 'OX1_STACORR', 'OX2',
       'OX2Q', 'OX2_CRUISECORR', 'OX2_STACORR', 'OXAVE_STACORR', 'OX1UM',
       'OX1UM_CRUISECORR', 'OX1UM_STACORR', 'OX2UM', 'OX2UM_CRUISECORR',
       'OX2UM_STACORR', 'OXAVEUM_STACORR', 'FLUORV', 'FLUORQ',
       'ESTCHL_CRUISECORR', 'ESTCHL_STACORR', 'ISUSV', 'ISUSQ',
       'ESTNO3_CRUISECORR', 'ESTNO3_STACORR', 'SIGTHETATS1', 'SIGTHETATS1Q',
       'SIGTHETATS2', 'SIGTHETATS2Q', 'BAT', 'XMISS', 'TRANSQ', 'PH', 'PHQ',
       'SPAR', 'SPARQ', 'PAR', 'PARQ', 'POT1', 'POT2', 'DYNHT', 'SVA',
       'OXSAT1', 'OXSAT2', 'BTL_DEPTH', 'BTL_TEMP', 'SALTB', 'OXB', 'OXBUM',
       'CHL_A', 'PHAEO', 'NO3', 'NO2', 

### CTD has CAST_COUNT and CAST_ID (not usable 8-digit string) while Secchi has Cst_Cnt (renamed to CAST_COUNT) and Cast_ID (with the usable long string that matches the CAST_ID from satellite.parquet)

In [ ]:
ctd_out = ctd.drop(columns=["ORD_OCC", "EVENT_NUM"])

# Coerce all mixed-type object columns to numeric where possible
for col in ctd_out.select_dtypes(include="object").columns:
    converted = pd.to_numeric(ctd_out[col], errors="coerce")
    if converted.notna().sum() > 0:  # only replace if column has numeric data
        ctd_out[col] = converted

ctd_out.to_parquet("../Data/Parquet/downcast_all.parquet", index=False)
print("Saved to ../Data/Parquet/downcast_all.parquet")

Saved to ../Data/Parquet/downcast_all.parquet


: 

In [6]:
import pandas as pd
secchi = pd.read_csv("../Data/1949-2021_Secchi.csv", low_memory=False)
secchi.head()

# for col in secchi.select_dtypes(include="object").columns:
#     converted = pd.to_numeric(secchi[col], errors="coerce")
#     if converted.notna().sum() > 0:
#         secchi[col] = converted

# secchi.to_parquet("../Data/Parquet/1949-2021_Secchi.parquet", index=False)
# print("Saved to ../Data/Parquet/1949-2021_Secchi.parquet")

,Cst_Cnt,Cruise_ID,Cruise,Cruz_Sta,DbSta_ID,Cast_ID,Sta_ID,Quarter,Sta_Code,Distance,...,Wave_Prd,Wind_Dir,Wind_Spd,Barometer,Dry_T,Wet_T,Wea,Cloud_Typ,Cloud_Amt,Visibility
0,1,1949-03-01-C-31CR,194903,1.949030e+13,5400560,19-4903CR-HY-060-0930-05400560,054.0 056.0,1,NST,NaN,...,NaN,23.0,18.0,NaN,NaN,NaN,2.0,NaN,NaN,NaN
1,2,1949-03-01-C-31CR,194903,1.949030e+13,5200750,19-4903CR-HY-060-2112-05200750,052.0 075.0,1,NST,NaN,...,NaN,16.0,5.0,NaN,NaN,NaN,4.0,NaN,NaN,NaN
2,3,1949-03-01-C-31CR,194903,1.949030e+13,5100850,19-4903CR-HY-061-0354-05100850,051.0 085.0,1,NST,NaN,...,NaN,23.0,5.0,NaN,NaN,NaN,6.0,NaN,NaN,NaN
3,4,1949-03-01-C-31CR,194903,1.949030e+13,5000950,19-4903CR-HY-061-1042-05000950,050.0 095.0,1,NST,NaN,...,NaN,18.0,8.0,NaN,NaN,NaN,2.0,NaN,NaN,NaN
4,5,1949-03-01-C-31CR,194903,1.949030e+13,5001040,19-4903CR-HY-061-1706-05001040,050.0 104.0,1,NST,NaN,...,NaN,27.0,13.0,NaN,NaN,NaN,7.0,NaN,NaN,NaN


## Run the following only once to merge datasets
### Keeps only CAST_COUNT's with missing Secchi

In [ ]:
downcast = pd.read_parquet("../Data/Parquet/downcast_all.parquet")
secchi = pd.read_parquet("../Data/Parquet/1949-2021_Secchi.parquet")

secchi = secchi.rename(columns={"Cst_Cnt": "CAST_COUNT"})

# Only keep CAST_COUNTs with missing Secchi — these are the ones to predict
secchi_missing = secchi[secchi["Secchi"].isna()]
print(f"CAST_COUNTs with missing Secchi: {secchi_missing['CAST_COUNT'].nunique()} / {secchi['CAST_COUNT'].nunique()} ({100*secchi_missing['CAST_COUNT'].nunique()/secchi['CAST_COUNT'].nunique():.1f}%)")

merged = downcast.merge(secchi_missing, on="CAST_COUNT", how="inner")

merged.to_parquet("../Data/Parquet/downcast_secchi_merged.parquet", index=False)
print(f"Merged shape: {merged.shape}")
print("Saved to ../Data/Parquet/downcast_secchi_merged.parquet")

CAST_COUNTs with missing Secchi: 29955 / 35644 (84.0%)
Merged shape: (2141218, 142)
Saved to ../Data/Parquet/downcast_secchi_merged.parquet


In [61]:
print(f"Merged unique cast counts: {merged['CAST_COUNT'].nunique()}")

Merged unique cast counts: 4980


## Run code from here for each session

In [1]:
import pandas as pd

merged = pd.read_parquet("../Data/Parquet/downcast_secchi_merged.parquet")
valid = merged[(merged["LAT_DEC"] != -99.0) & (merged["LON_DEC"] != -99.0)]
print(valid["CAST_COUNT"].nunique())

4977


In [2]:
# How many rows are fully valid across both coordinates
valid_rows = merged[(merged["LAT_DEC"] != -99.0) & (merged["LON_DEC"] != -99.0)]
print(len(merged) - len(valid_rows))  # Total invalid rows

# Which casts are partially bad (have some invalid rows but aren't entirely bad)
cast_invalid_counts = merged[
    (merged["LAT_DEC"] == -99.0) & (merged["LON_DEC"] == -99.0)
].groupby("CAST_COUNT").size()
print(f"Casts with at least one invalid row: {len(cast_invalid_counts)}")

334827
Casts with at least one invalid row: 721


### Format of this merged dataset
Need to aggregate the CTD variables so that there is 1 value per cast count and thus one response (Secchi value) too
- Create the photic depth threshold again - 2m baseline with 4.7% light threshold
- Aggregate XMiss, CHL-A, ESTCHL, PHAEO, (sum to min{photic_depth, median_secchi=20}), calc KPAR and KPAR_SLOPE 
- Select cast counts to predict secchi for times within the 1 hr after sunrise 1 hour before sunset interval, by using real sunset/sunrise times for the specific date_time

In [3]:
# replace -99 LAT_DEC and LON_DEC values with NaN 
merged["LAT_DEC"] = merged["LAT_DEC"].replace(-99.0, float("nan"))
merged["LON_DEC"] = merged["LON_DEC"].replace(-99.0, float("nan"))

print(merged[["LAT_DEC", "LON_DEC"]].isna().sum()) # print missing values


LAT_DEC    377230
LON_DEC    378198
dtype: int64


#### Lat and Lon coords should all be the same within a Cast, so if some rows within a specific Cast are nan, replace them with the valid lat/lon coordinates

In [4]:
merged[["LAT_DEC", "LON_DEC"]] = merged.groupby("CAST_COUNT")[["LAT_DEC", "LON_DEC"]].transform("first")

In [5]:
# now missing lat/lon values pertain to casts that don't have any valid lat/lon rows
print(merged[["LAT_DEC", "LON_DEC"]].isna().sum())

LAT_DEC    377230
LON_DEC    377230
dtype: int64


In [ ]:
fully_unresolvable = merged[merged["LAT_DEC"].isna()]["CAST_COUNT"].nunique()
print(f"Casts with no valid LAT at all: {fully_unresolvable}")

Casts with no valid LAT at all: 816


### Drop rows that have missing lat/lon

In [7]:
valid_casts = merged.dropna(subset=["LAT_DEC", "LON_DEC"])

print(f"Casts before: {merged['CAST_COUNT'].nunique()}")        # 4980
print(f"Casts after:  {valid_casts['CAST_COUNT'].nunique()}")   # expect 4980 - 816 = 4164
print(f"Rows dropped: {len(merged) - len(valid_casts)}")        # expect ~377,230
print(f"Rows remaining: {len(valid_casts)}")                   # expect ~1,000,000 - 377,230 = ~622,770

Casts before: 4980
Casts after:  4164
Rows dropped: 377230
Rows remaining: 1763988


### valid_casts object should not contain any NaN lat/lon values

In [12]:
# does valid_casts contain any NaN lat values or any Nan lon values?
print(valid_casts[["LAT_DEC", "LON_DEC"]].isna().sum())

LAT_DEC    0
LON_DEC    0
dtype: int64


Note: valid_casts, the object with all NaN coords removed  is not written to parquet yet

The Latitude and Longitude coordinates of each station are very similar, and the CCE is a relatively small region likely to be unaffected by taking an average of coordinates to use to calculate sunrise/sunset times.

Thus, I'll use a single average of each coordinate.

In [14]:
# Average across all rows with valid coordinates
avg_lat = valid_casts["LAT_DEC"].mean()
avg_lon = valid_casts["LON_DEC"].mean()
print(f"Average LAT: {avg_lat:.4f}")
print(f"Average LON: {avg_lon:.4f}")

Average LAT: 32.9021
Average LON: -120.3516


So we'll use the average `LAT` and `LON` along with `DATE_TIME_UTC` to find the daylight window for each cast.

In [4]:
!pip install astral -q

In [16]:
from astral import LocationInfo
from astral.sun import sun
from datetime import timezone, timedelta
from zoneinfo import ZoneInfo

LA_TZ = ZoneInfo("America/Los_Angeles")

df = valid_casts[["CAST_COUNT", "DATE_TIME_UTC"]].copy()
df["DATE_TIME_UTC"] = pd.to_datetime(df["DATE_TIME_UTC"], utc=True)
df = df.dropna(subset=["DATE_TIME_UTC"])

# Extract local date using America/Los_Angeles (accounts for DST automatically)
df["local_date"] = df["DATE_TIME_UTC"].dt.tz_convert(LA_TZ).dt.date

# Single location using average coordinates
loc = LocationInfo(latitude=avg_lat, longitude=avg_lon)

# Compute sunrise/sunset once per unique local date using LA timezone
date_cache = {}
for d in df["local_date"].unique():
    try:
        s = sun(loc.observer, date=d, tzinfo=LA_TZ)
        date_cache[d] = (s["sunrise"] + timedelta(hours=1), s["sunset"] - timedelta(hours=1))
    except Exception:
        date_cache[d] = (None, None)

df["window_start"] = df["local_date"].map(lambda d: date_cache.get(d, (None, None))[0])
df["window_end"]   = df["local_date"].map(lambda d: date_cache.get(d, (None, None))[1])

mask = (
    df["window_start"].notna() &
    df["window_end"].notna() &
    (df["DATE_TIME_UTC"] >= df["window_start"]) &
    (df["DATE_TIME_UTC"] <= df["window_end"])
)

daylight_cast_counts = df.loc[mask, "CAST_COUNT"].unique()
daylight_cast_set = set(daylight_cast_counts)
print(f"CAST_COUNT values within daylight window: {len(daylight_cast_counts)}")
print(daylight_cast_counts[:20]) # print tail of cast_counts in daylight window

CAST_COUNT values within daylight window: 986
[28868. 28869. 28873. 28874. 28877. 28881. 28884. 28887. 28888. 28893.
 28894. 28898. 28899. 28902. 28903. 28906. 28909. 28910. 28915. 28919.]


In [17]:
from datetime import timezone, timedelta, date as date_type

LA_TZ = ZoneInfo("America/Los_Angeles")
loc_test = LocationInfo(latitude=avg_lat, longitude=avg_lon)
s_test = sun(loc_test.observer, date=date_type(2000, 8, 11), tzinfo=LA_TZ)
window_start = s_test["sunrise"] + timedelta(hours=1)
window_end   = s_test["sunset"]  - timedelta(hours=1)
print(f"Sunrise+1h (UTC):   {window_start.astimezone(timezone.utc)}")
print(f"Sunset-1h  (UTC):   {window_end.astimezone(timezone.utc)}")
print(f"Sunrise+1h (local): {window_start}")
print(f"Sunset-1h  (local): {window_end}")
print(f"Window valid (start < end): {window_start < window_end}")

Sunrise+1h (UTC):   2000-08-11 14:22:25.348828+00:00
Sunset-1h  (UTC):   2000-08-12 01:50:03.084478+00:00
Sunrise+1h (local): 2000-08-11 07:22:25.348828-07:00
Sunset-1h  (local): 2000-08-11 18:50:03.084478-07:00
Window valid (start < end): True


In [18]:
# Rows with invalid coordinates (NaN) - should be 0 after filtering
invalid_coord_mask = (
    valid_casts["LAT_DEC"].isna() |
    valid_casts["LON_DEC"].isna()
)

invalid_coords_df = valid_casts[invalid_coord_mask]

# Of those, which CAST_COUNTs were identified as within the daylight window
# (inferred from other rows of the same cast with valid coordinates)
daylight_cast_set = set(daylight_cast_counts)
inferred_daylight = invalid_coords_df[invalid_coords_df["CAST_COUNT"].isin(daylight_cast_set)]

inferred_cast_counts = inferred_daylight["CAST_COUNT"].unique()
print(f"CAST_COUNT values with invalid coords but inferred to be in daylight window: {len(inferred_cast_counts)}")
print(inferred_cast_counts[:20])

CAST_COUNT values with invalid coords but inferred to be in daylight window: 0
[]


### Now the valid_casts object should be filtered to contain CAST_COUNT's only in the daylight_cast_counts (daylight window)
## Renamed: df_slim

In [ ]:
df_slim = valid_casts[valid_casts["CAST_COUNT"].isin(daylight_cast_counts)]
print(f"Remaining CAST_COUNTs: {valid_casts['CAST_COUNT'].nunique()}")
print(f"Remaining rows: {len(valid_casts)}")

Remaining CAST_COUNTs: 986
Remaining rows: 379367


### Keep df with necessary 14 columns only, before feature engineering

#### Keep: Cruise_ID, Cruise, Sta_ID (for map visualization)

In [33]:
cols = ["Cast_ID", "DATE_TIME_UTC", "Cruise_ID", "Cruise", "Sta_ID", "DATE_TIME_PST", "LAT_DEC", "LON_DEC", "DEPTH", "ESTCHL_STACORR", "XMISS", "PAR", "CAST_COUNT", "Secchi"]

In [34]:
df_slim = df_slim[cols].copy()
print(f"Shape: {df_slim.shape}")
print(df_slim.head())

Shape: (379367, 14)
                               Cast_ID         DATE_TIME_UTC  \
377230  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:46Z   
377231  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:58Z   
377232  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:59Z   
377233  19-9807NM-MX-191-0414-09330350  1998-07-08T20:01:01Z   
377234  19-9807NM-MX-191-0414-09330350  1998-07-08T20:01:04Z   

                Cruise_ID  Cruise       Sta_ID         DATE_TIME_PST  LAT_DEC  \
377230  1998-07-09-C-32NM  199807  093.3 035.0  1998-07-08T12:00:46Z    32.68   
377231  1998-07-09-C-32NM  199807  093.3 035.0  1998-07-08T12:00:58Z    32.68   
377232  1998-07-09-C-32NM  199807  093.3 035.0  1998-07-08T12:00:59Z    32.68   
377233  1998-07-09-C-32NM  199807  093.3 035.0  1998-07-08T12:01:01Z    32.68   
377234  1998-07-09-C-32NM  199807  093.3 035.0  1998-07-08T12:01:04Z    32.68   

        LON_DEC  DEPTH  ESTCHL_STACORR  XMISS   PAR  CAST_COUNT  Secchi  
377230  -117.87    1.0            

In [35]:
# Minimum depth per cast (i.e. the shallowest measurement = starting depth)
min_depth_per_cast = df_slim.groupby("CAST_COUNT")["DEPTH"].min()

# Round to nearest integer and count how many casts start at each depth
starting_depth_counts = min_depth_per_cast.round(0).astype(int).value_counts().sort_index()
print("Starting depth (m) : # of CAST_COUNTs")
print(starting_depth_counts.to_string())

Starting depth (m) : # of CAST_COUNTs
DEPTH
1     356
2     502
3     107
4      14
5       3
6       2
8       1
12      1


In [36]:
casts_with_2m = df_slim[df_slim["DEPTH"].round(0) == 2]["CAST_COUNT"].nunique()
print(f"Unique CAST_COUNTs with a 2m depth measurement: {casts_with_2m}")

Unique CAST_COUNTs with a 2m depth measurement: 858


In [37]:
non_integer_depths = df_slim[df_slim["DEPTH"] != df_slim["DEPTH"].round(0)]
print(f"Rows with non-integer DEPTH: {len(non_integer_depths)}")
print(non_integer_depths["DEPTH"].describe())
print(non_integer_depths["DEPTH"].head(20))

Rows with non-integer DEPTH: 0
count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: DEPTH, dtype: float64
Series([], Name: DEPTH, dtype: float64)


### Drop NaN PAR value rows

In [38]:
df_slim = df_slim.dropna(subset=["PAR"])
print(f"Rows after dropping NaN PAR: {len(df_slim)}")  # expect 379367 - 22444 = 356923
print(f"Casts remaining: {df_slim['CAST_COUNT'].nunique()}")  # expect 986 - 64 = 922

Rows after dropping NaN PAR: 356923
Casts remaining: 922


#### Some more CAST_COUNT's will be dropped (one's that don't contain 2m depth rows)

In [39]:
# Only keep CAST_COUNTs that have a measurement at exactly 2m
casts_with_2m = df_slim[df_slim["DEPTH"].round(0) == 2]["CAST_COUNT"].unique()
df_photic = df_slim[df_slim["CAST_COUNT"].isin(casts_with_2m)].copy()

# Get PAR at 2m baseline for each cast
par_at_2m = (
    df_photic[df_photic["DEPTH"] == 2]
    .groupby("CAST_COUNT")["PAR"]
    .first()
    .rename("PAR_2m")
)
df_photic = df_photic.merge(par_at_2m, on="CAST_COUNT", how="left")

# Calculate photic zone threshold (4.7% of PAR at 2m)
df_photic["PAR_THRESHOLD"] = df_photic["PAR_2m"] * 0.047

# For each cast, find the row whose PAR is closest to the threshold
df_photic["PAR_DIFF"] = (df_photic["PAR"] - df_photic["PAR_THRESHOLD"]).abs()
photic_idx = df_photic.groupby("CAST_COUNT")["PAR_DIFF"].idxmin()

df_photic["PHOTIC_ZONE"] = False
df_photic.loc[photic_idx, "PHOTIC_ZONE"] = True

print(f"CAST_COUNTs with photic zone identified: {df_photic["PHOTIC_ZONE"].sum()}")
print(df_photic[df_photic["PHOTIC_ZONE"]][["CAST_COUNT", "DEPTH", "PAR", "PAR_2m", "PAR_THRESHOLD"]].head(10))

CAST_COUNTs with photic zone identified: 779
      CAST_COUNT  DEPTH   PAR  PAR_2m  PAR_THRESHOLD
0        28868.0    1.0  0.64    0.64        0.03008
512      28869.0    1.0  0.64    0.64        0.03008
1033     28874.0    1.0  0.64    0.64        0.03008
1546     28877.0    2.0  0.64    0.64        0.03008
2065     28881.0    2.0  0.64    0.64        0.03008
2586     28884.0    2.0  0.64    0.64        0.03008
3110     28887.0    1.0  0.64    0.64        0.03008
3630     28888.0    1.0  0.64    0.64        0.03008
4152     28893.0    1.0  0.64    0.64        0.03008
4202     28894.0    1.0  0.64    0.64        0.03008


In [40]:
print(df_photic.head())

                          Cast_ID         DATE_TIME_UTC          Cruise_ID  \
0  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:46Z  1998-07-09-C-32NM   
1  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:58Z  1998-07-09-C-32NM   
2  19-9807NM-MX-191-0414-09330350  1998-07-08T20:00:59Z  1998-07-09-C-32NM   
3  19-9807NM-MX-191-0414-09330350  1998-07-08T20:01:01Z  1998-07-09-C-32NM   
4  19-9807NM-MX-191-0414-09330350  1998-07-08T20:01:04Z  1998-07-09-C-32NM   

   Cruise       Sta_ID         DATE_TIME_PST  LAT_DEC  LON_DEC  DEPTH  \
0  199807  093.3 035.0  1998-07-08T12:00:46Z    32.68  -117.87    1.0   
1  199807  093.3 035.0  1998-07-08T12:00:58Z    32.68  -117.87    2.0   
2  199807  093.3 035.0  1998-07-08T12:00:59Z    32.68  -117.87    3.0   
3  199807  093.3 035.0  1998-07-08T12:01:01Z    32.68  -117.87    4.0   
4  199807  093.3 035.0  1998-07-08T12:01:04Z    32.68  -117.87    5.0   

   ESTCHL_STACORR  XMISS   PAR  CAST_COUNT  Secchi  PAR_2m  PAR_THRESHOLD  \
0            0.

### `K_PAR` and `K_PAR_SLOPE` calculations (and NaNs check)

In [41]:
import numpy as np
from scipy.stats import linregress

# Get photic zone depth per cast (row where PHOTIC_ZONE == True)
photic_rows = df_photic[df_photic["PHOTIC_ZONE"]][["CAST_COUNT", "DEPTH", "PAR"]].rename(
    columns={"DEPTH": "PHOTIC_DEPTH", "PAR": "PAR_z"}
)

# Merge PAR_2m into photic_rows
photic_rows = photic_rows.merge(
    df_photic[df_photic["DEPTH"] == 2].groupby("CAST_COUNT")["PAR"].first().rename("PAR_2m"),
    on="CAST_COUNT", how="left"
)

# K_PAR = -ln(PAR(z) / PAR(2m)) / z
photic_rows["K_PAR"] = -np.log(photic_rows["PAR_z"] / photic_rows["PAR_2m"]) / photic_rows["PHOTIC_DEPTH"]

# K_PAR_SLOPE: linear fit of ln(PAR) vs depth over [2m, min(photic_depth, 20m)]
def calc_kpar_slope(cast_count, photic_depth):
    window_max = min(photic_depth, 20)
    rows = df_photic[
        (df_photic["CAST_COUNT"] == cast_count) &
        (df_photic["DEPTH"] >= 2) &
        (df_photic["DEPTH"] <= window_max) &
        (df_photic["PAR"] > 0)
    ][["DEPTH", "PAR"]].dropna()
    if len(rows) < 2 or rows["DEPTH"].nunique() < 2:
        return np.nan
    slope, _, _, _, _ = linregress(rows["DEPTH"], np.log(rows["PAR"]))
    return -slope

photic_rows["K_PAR_SLOPE"] = photic_rows.apply(
    lambda r: calc_kpar_slope(r["CAST_COUNT"], r["PHOTIC_DEPTH"]), axis=1
)

print(photic_rows[["CAST_COUNT", "PHOTIC_DEPTH", "PAR_2m", "PAR_z", "K_PAR", "K_PAR_SLOPE"]].head(10))
print(f"K_PAR NaNs: " + str(photic_rows["K_PAR"].isna().sum()))
print(f"K_PAR_SLOPE NaNs: " + str(photic_rows["K_PAR_SLOPE"].isna().sum()))

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: invalid value encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


   CAST_COUNT  PHOTIC_DEPTH  PAR_2m  PAR_z  K_PAR  K_PAR_SLOPE
0     28868.0           1.0    0.64   0.64   -0.0          NaN
1     28869.0           1.0    0.64   0.64   -0.0          NaN
2     28874.0           1.0    0.64   0.64   -0.0          NaN
3     28877.0           2.0    0.64   0.64   -0.0          NaN
4     28881.0           2.0    0.64   0.64   -0.0          NaN
5     28884.0           2.0    0.64   0.64   -0.0          NaN
6     28887.0           1.0    0.64   0.64   -0.0          NaN
7     28888.0           1.0    0.64   0.64   -0.0          NaN
8     28893.0           1.0    0.64   0.64   -0.0          NaN
9     28894.0           1.0    0.64   0.64   -0.0          NaN
K_PAR NaNs: 19
K_PAR_SLOPE NaNs: 216


In [42]:
# --- K_PAR NaN diagnosis ---
kpar_nan = photic_rows[photic_rows["K_PAR"].isna()][["CAST_COUNT", "PHOTIC_DEPTH", "PAR_z", "PAR_2m"]]
print(f"K_PAR NaN rows ({len(kpar_nan)}):", kpar_nan.to_string())

print("--- K_PAR_SLOPE NaN diagnosis ---")
kpar_slope_nan_casts = photic_rows[photic_rows["K_PAR_SLOPE"].isna()]["CAST_COUNT"].values

reasons = {"PAR_missing_or_zero": 0, "fewer_than_2_rows": 0, "all_same_depth": 0}
for cast_count in kpar_slope_nan_casts:
    photic_depth = photic_rows.loc[photic_rows["CAST_COUNT"] == cast_count, "PHOTIC_DEPTH"].values[0]
    window_max = min(photic_depth, 20)
    rows = df_photic[
        (df_photic["CAST_COUNT"] == cast_count) &
        (df_photic["DEPTH"] >= 2) &
        (df_photic["DEPTH"] <= window_max) &
        (df_photic["PAR"] > 0)
    ][["DEPTH", "PAR"]].dropna()
    if len(rows) < 2:
        reasons["fewer_than_2_rows"] += 1
    elif rows["DEPTH"].nunique() < 2:
        reasons["all_same_depth"] += 1
    else:
        reasons["PAR_missing_or_zero"] += 1

print("Reason breakdown:", reasons)

# Sample window for first 3 slope-NaN casts
print("Sample window contents for first 3 K_PAR_SLOPE NaN casts:")
for cast_count in kpar_slope_nan_casts[:3]:
    photic_depth = photic_rows.loc[photic_rows["CAST_COUNT"] == cast_count, "PHOTIC_DEPTH"].values[0]
    window_max = min(photic_depth, 20)
    rows = df_photic[
        (df_photic["CAST_COUNT"] == cast_count) &
        (df_photic["DEPTH"] >= 2) &
        (df_photic["DEPTH"] <= window_max)
    ][["DEPTH", "PAR"]]
    print(f"CAST_COUNT={cast_count}, photic_depth={photic_depth}, window=[2, {window_max}]")
    print(rows.to_string())


K_PAR NaN rows (19):      CAST_COUNT  PHOTIC_DEPTH         PAR_z  PAR_2m
125     29281.0           3.0 -6.200000e+01  351.45
231     30784.0          38.0 -9.990000e-29    2.10
232     30785.0          26.0 -9.990000e-29    2.00
234     30790.0          50.0 -9.990000e-29    2.07
235     30794.0          36.0 -9.990000e-29    1.85
236     30780.0          39.0 -9.990000e-29    1.69
237     30776.0          40.0 -9.990000e-29    1.95
240     30755.0          38.0 -9.990000e-29    2.25
241     30760.0          32.0 -9.990000e-29    2.03
242     30764.0          35.0 -9.990000e-29    1.91
243     30749.0          33.0 -9.990000e-29    1.79
245     30744.0          41.0 -9.990000e-29    1.80
247     30732.0          31.0 -9.990000e-29    1.80
248     30733.0          31.0 -9.990000e-29    1.92
249     30737.0          31.0 -9.990000e-29    2.05
250     30728.0          33.0 -9.990000e-29    1.73
251     30724.0          25.0 -9.990000e-29    2.07
252     30723.0          39.0 -9.990000e-29

In [43]:
total = len(photic_rows)
after_kpar = photic_rows[photic_rows["K_PAR"].notna()]
after_both = after_kpar[after_kpar["K_PAR_SLOPE"].notna()]
print(f"Total CAST_COUNTs: {total}")
print(f"After dropping K_PAR NaNs: {len(after_kpar)} (lost {total - len(after_kpar)})")
print(f"After also dropping K_PAR_SLOPE NaNs: {len(after_both)} (lost {len(after_kpar) - len(after_both)})")
print(f"Final usable CAST_COUNTs: {len(after_both)} / {total} ({100*len(after_both)/total:.1f}%)")


Total CAST_COUNTs: 779
After dropping K_PAR NaNs: 760 (lost 19)
After also dropping K_PAR_SLOPE NaNs: 545 (lost 215)
Final usable CAST_COUNTs: 545 / 779 (70.0%)


In [44]:
# Keep only casts with valid K_PAR and K_PAR_SLOPE
photic_rows = photic_rows[photic_rows["K_PAR"].notna() & photic_rows["K_PAR_SLOPE"].notna()].reset_index(drop=True)
print(f"Retained CAST_COUNTs: {len(photic_rows)}")


Retained CAST_COUNTs: 545


### So, after filtering the K_PAR/slope values for NaNs, we still have 545 Casts usable

### Now build integration window to sum up XMISS and ESTCHL

In [45]:
# Build integration window per cast: [2m, min(photic_depth, 20m)]
photic_depth_map = photic_rows.set_index("CAST_COUNT")["PHOTIC_DEPTH"]

def sum_over_window(cast_count, col):
    window_max = min(photic_depth_map[cast_count], 20)
    rows = df_photic[
        (df_photic["CAST_COUNT"] == cast_count) &
        (df_photic["DEPTH"] >= 2) &
        (df_photic["DEPTH"] <= window_max)
    ][col]
    return rows.sum(min_count=1)

cast_counts = photic_rows["CAST_COUNT"].values

photic_rows["XMISS_SUMMED"]  = [sum_over_window(c, "XMISS")          for c in cast_counts]
photic_rows["ESTCHL_SUMMED"] = [sum_over_window(c, "ESTCHL_STACORR") for c in cast_counts]

print(photic_rows[["CAST_COUNT", "XMISS_SUMMED", "ESTCHL_SUMMED"]].head(10))
print("NaN counts:")
print(photic_rows[["XMISS_SUMMED", "ESTCHL_SUMMED"]].isna().sum())


   CAST_COUNT  XMISS_SUMMED  ESTCHL_SUMMED
0     29275.0        798.45           1.71
1     29276.0        801.46           1.53
2     29277.0        795.90           3.08
3     29282.0         86.56           0.18
4     29285.0        809.93           0.87
5     29292.0         84.87           0.12
6     29293.0        802.06           1.71
7     29296.0        776.98           6.58
8     29380.0        798.59           1.37
9     29384.0        787.90          12.96
NaN counts:
XMISS_SUMMED      0
ESTCHL_SUMMED    10
dtype: int64


### `CHL_A_SUMMED` & `PHAEO_SUMMED` are very sparse which is why they weren't great predictors in prior analyses, so 

In [ ]:
# Drop the 10 casts with missing ESTCHL_SUMMED
photic_rows = photic_rows[photic_rows["ESTCHL_SUMMED"].notna()].reset_index(drop=True)
print(f"Retained CAST_COUNTs after ESTCHL_SUMMED filter: {len(photic_rows)}")


Retained CAST_COUNTs after ESTCHL_SUMMED filter: 535


In [47]:
print(photic_rows.head())

   CAST_COUNT  PHOTIC_DEPTH   PAR_z   PAR_2m     K_PAR  K_PAR_SLOPE  \
0     29275.0          54.0  5529.1   9999.0  0.010971    -0.000000   
1     29276.0          41.0  6025.7   9283.1  0.010541    -0.001163   
2     29277.0          41.0  6664.6  10017.0  0.009938     0.000028   
3     29282.0           3.0  2527.5   4116.6  0.162599     0.487797   
4     29285.0          43.0  8066.9   9959.1  0.004900    -0.000064   

   XMISS_SUMMED  ESTCHL_SUMMED  
0        798.45           1.71  
1        801.46           1.53  
2        795.90           3.08  
3         86.56           0.18  
4        809.93           0.87  


### Add photic_rows object to final df

In [48]:
meta_cols = ["CAST_COUNT", "Cast_ID", "Cruise_ID", "Cruise", "Sta_ID",
             "DATE_TIME_UTC", "DATE_TIME_PST", "LAT_DEC", "LON_DEC"]
cast_meta = df_photic.groupby("CAST_COUNT")[meta_cols[1:]].first().reset_index()

df_final = (
    photic_rows[["CAST_COUNT", "PHOTIC_DEPTH", "K_PAR", "K_PAR_SLOPE", "ESTCHL_SUMMED", "XMISS_SUMMED"]]
    .rename(columns={"PHOTIC_DEPTH": "DEPTH"})
    .merge(cast_meta, on="CAST_COUNT", how="left")
)

df_final = df_final[["CAST_COUNT", "Cast_ID", "Cruise_ID", "Cruise", "Sta_ID",
                      "DATE_TIME_UTC", "DATE_TIME_PST", "LAT_DEC", "LON_DEC",
                      "DEPTH", "K_PAR", "K_PAR_SLOPE", "ESTCHL_SUMMED", "XMISS_SUMMED"]]

print(f"Shape: {df_final.shape}")
print(f"Missing values:\n{df_final.isna().sum()}")
print(df_final.head())

Shape: (535, 14)
Missing values:
CAST_COUNT       0
Cast_ID          0
Cruise_ID        0
Cruise           0
Sta_ID           0
DATE_TIME_UTC    0
DATE_TIME_PST    0
LAT_DEC          0
LON_DEC          0
DEPTH            0
K_PAR            0
K_PAR_SLOPE      0
ESTCHL_SUMMED    0
XMISS_SUMMED     0
dtype: int64
   CAST_COUNT                         Cast_ID          Cruise_ID  Cruise  \
0     29275.0  19-9910NM-MX-277-0107-09330300  1999-10-03-C-32NM  199910   
1     29276.0  19-9910NM-MX-277-0502-09330350  1999-10-03-C-32NM  199910   
2     29277.0  19-9910NM-MX-277-0856-09330400  1999-10-03-C-32NM  199910   
3     29282.0  19-9910NM-MX-278-0811-09330700  1999-10-03-C-32NM  199910   
4     29285.0  19-9910NM-MX-279-0511-09331000  1999-10-03-C-32NM  199910   

        Sta_ID         DATE_TIME_UTC         DATE_TIME_PST  LAT_DEC  LON_DEC  \
0  093.3 030.0  1999-10-03T16:51:45Z  1999-10-03T08:51:45Z    32.84  -117.53   
1  093.3 035.0  1999-10-03T20:49:44Z  1999-10-03T12:49:44Z    32.67  -1

In [49]:
df_final.to_parquet("../Data/Parquet/daylight_stations_to_predict.parquet", index=False)
print(f"Saved {len(df_final)} rows to ../Data/Parquet/daylight_stations_to_predict.parquet")


Saved 535 rows to ../Data/Parquet/daylight_stations_to_predict.parquet


## Wrote the final df to parquet: `daylight_stations_to_predict.parquet`

In [54]:
daylight = pd.read_parquet("../Data/Parquet/daylight_stations_to_predict.parquet")
print(daylight.head(15))

    CAST_COUNT                         Cast_ID          Cruise_ID  Cruise  \
0      29275.0  19-9910NM-MX-277-0107-09330300  1999-10-03-C-32NM  199910   
1      29276.0  19-9910NM-MX-277-0502-09330350  1999-10-03-C-32NM  199910   
2      29277.0  19-9910NM-MX-277-0856-09330400  1999-10-03-C-32NM  199910   
3      29282.0  19-9910NM-MX-278-0811-09330700  1999-10-03-C-32NM  199910   
4      29285.0  19-9910NM-MX-279-0511-09331000  1999-10-03-C-32NM  199910   
5      29292.0  19-9910NM-MX-281-0212-09000800  1999-10-03-C-32NM  199910   
6      29293.0  19-9910NM-MX-281-0751-09000700  1999-10-03-C-32NM  199910   
7      29296.0  19-9910NM-MX-282-0412-09000450  1999-10-03-C-32NM  199910   
8      29380.0  20-0001NM-MX-018-0513-08331000  2000-01-07-C-32NM  200001   
9      29384.0  20-0001NM-MX-019-0433-08330600  2000-01-07-C-32NM  200001   
10     29389.0  20-0001NM-MX-020-0126-08180469  2000-01-07-C-32NM  200001   
11     29393.0  20-0001NM-MX-021-0349-08000700  2000-01-07-C-32NM  200001   

In [53]:
import pandas as pd

satellite = pd.read_parquet("../Data/Parquet/satellite.parquet")
satellite.head(15)

,Lon_Dec,Lat_Dec,Date,Time,Cruise,St_Line,Cast_ID,Sta_ID,Secchi,IntChl,...,SceneCenterSolarZenith,VarName,Pointvalue,Nin,Nout,Min,Max,Mean,StDev,Median
0,-120.01,33.15333333,9/27/1997,22:55,199709,86.7,19-9709NM-MX-270-2255-08670550,086.7 055.0,16,48.5,...,-999,Zsd,17.7,9,0,16.756,19.116,17.83111,0.843606,17.464
1,-120.815,34.33166667,10/2/1997,18:03,199709,80,19-9709NM-PR-275-1803-08000550,080.0 055.0,15,61.20000076,...,-999,Zsd,14.868,9,0,14.16,14.868,14.55333,0.263856,14.632
2,-118.3783333,33.18166667,4/8/1998,20:43,199804,90,19-9804JD-MX-098-2043-09000370,090.0 037.0,18,45.70000076,...,-999,Zsd,15.104,9,0,14.16,15.576,14.76311,0.4587016,14.632
3,-118.4916667,33.89,4/9/1998,15:16,199804,86.7,19-9804JD-MX-099-1516-08670330,086.7 033.0,7,196.6000061,...,-999,Zsd,7.552,9,0,5.9,7.788,7.106222,0.6925348,7.552
4,-118.6283333,33.82333333,4/9/1998,18:48,199804,86.7,19-9804JD-PR-099-1848-08670350,086.7 035.0,10,188,...,-999,Zsd,9.912,9,0,7.788,9.912,9.151556,0.8891422,9.44
5,-120.0233333,34.275,4/14/1998,23:35,199804,81.8,19-9804JD-MX-104-2335-08180469,081.8 046.9,4,361.8999939,...,-999,Zsd,4.248,9,0,4.012,4.484,4.195556,0.1573333,4.248
6,-121.6066667,35.47666667,4/18/1998,18:43,199804,73,19-9804JD-PR-108-1843-07330550,073.3 055.0,12,78.40000153,...,-999,Zsd,9.204,9,0,8.496,11.328,9.125334,0.8830311,8.968
7,-123.13,36.28666667,4/21/1998,23:08,199804,66.7,19-9804JD-MX-111-2308-06670650,066.7 065.0,18,45.79999924,...,-999,Zsd,21.948,9,0,21.476,22.892,22.23644,0.4975317,22.42
8,-119.51,34.17833333,9/24/1998,17:55,199809,83.3,19-9809NM-PR-267-1755-08330420,083.3 042.0,18,52.29999924,...,-999,Zsd,12.98,9,0,12.036,12.98,12.63911,0.2917038,12.744
9,-120.0183333,34.27333333,9/24/1998,21:08,199809,81.8,19-9809NM-MX-267-2108-08180469,081.8 046.9,16,45.20000076,...,-999,Zsd,16.048,9,0,14.632,16.048,15.49733,0.6016843,15.576


### Merge satellite.parquet so that we can validate predicted Secchi with satellite estimates (Pointvalue)
### Use Cast_ID to merge the dataframes

In [52]:
daylight = pd.read_parquet("../Data/Parquet/daylight_stations_to_predict.parquet")

daylight_cast_ids = set(daylight["Cast_ID"])
satellite_cast_ids = set(satellite["Cast_ID"])

common = daylight_cast_ids & satellite_cast_ids
print(f"Cast_IDs in daylight:             {len(daylight_cast_ids)}")
print(f"Cast_IDs in satellite:            {len(satellite_cast_ids)}")
print(f"Cast_IDs in both (intersection):  {len(common)}")
print()
for cast_id in sorted(common):
    print(cast_id)

Cast_IDs in daylight:             535
Cast_IDs in satellite:            443
Cast_IDs in both (intersection):  0



## UPDATE: daylight dataset has no Cast_ID's in common with the satellite dataset

### No merge can be performed